browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,277 @@
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() {
+  net::NetworkChangeNotifier::RemoveDNSObserver(this);
+}
+
+void BraveVpnDnsObserverService::OnDNSChanged() {
+  // Re-create service for reading config once becasue the service keeps
+  // the callback and cannot read multiple times.
+  dns_config_service_ = net::DnsConfigService::CreateSystemService();
+  dns_config_service_->ReadConfig(
+      base::BindRepeating(&BraveVpnDnsObserverService::OnSystemDNSConfigChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+void BraveVpnDnsObserverService::OnSystemDNSConfigChanged(
+    const net::DnsConfig& config) {
+  system_dns_config_ = config;
+  dns_config_service_.reset();
+}
+
+bool BraveVpnDnsObserverService::IsDNSSecure(
+    SecureDnsConfig browser_dns_config) const {
+  bool secure = (SecureDnsConfig::ModeToString(browser_dns_config.mode()) ==
+                 SecureDnsConfig::kModeSecure);
+  bool is_valid_automatic_mode =
+      (SecureDnsConfig::ModeToString(browser_dns_config.mode()) ==
+       SecureDnsConfig::kModeAutomatic) &&
+      IsValidDoHTemplates(browser_dns_config.doh_servers().ToString());
+  bool is_system_doh_enabled =
+      (!system_dns_config_.doh_config.servers().empty() &&
+       system_dns_config_.allow_dns_over_https_upgrade);
+  return secure || is_valid_automatic_mode || is_system_doh_enabled;
+}
+
+bool BraveVpnDnsObserverService::ShouldAllowExternalChanges() const {
+  if (allow_changes_for_testing_.has_value())
+    return allow_changes_for_testing_.value();
+
+  return (chrome::ShowQuestionMessageBoxSync(
+              GetAnchorBrowserWindow(),
+              l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+              l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_CHANGE_ALERT)) ==
+          chrome::MESSAGE_BOX_RESULT_YES);
+}
+
+bool BraveVpnDnsObserverService::IsDnsModeConfiguredByPolicy() const {
+  bool doh_policy_set =
+      policy_reader_ &&
+      !policy_reader_.Run(policy::key::kDnsOverHttpsMode).empty();
+  return doh_policy_set ||
+         SystemNetworkContextManager::GetStubResolverConfigReader()
+             ->ShouldDisableDohForManaged();
+}
+
+void BraveVpnDnsObserverService::ShowPolicyWarningMessage() {
+  if (!profile_prefs_->GetBoolean(prefs::kBraveVPNShowDNSPolicyWarningDialog)) {
+    return;
+  }
+
+  if (policy_callback_) {
+    std::move(policy_callback_).Run();
+    return;
+  }
+
+  chrome::ShowWarningMessageBoxWithCheckbox(
+      GetAnchorBrowserWindow(), l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_ALERT),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_CHECKBOX),
+      base::BindOnce(&SkipDNSDialog, profile_prefs_));
+}
+
+void BraveVpnDnsObserverService::ShowMessageWhyWeOverrideDNSSettings() {
+  if (skip_notification_dialog_for_testing_)
+    return;
+  VpnNotificationDialogView::Show(chrome::FindLastActive());
+}
+
+void BraveVpnDnsObserverService::OnDNSPrefChanged() {
```

Comment:
the prefs setting is irrelevant and we should not be changing it

Reply:
also it should be locked anyway because we set it by managed prefs

Reply:
done